# Неделя 1. Решение на R


In [3]:
df <- read.csv(file="mlbootcamp5_train.csv", header=TRUE, sep=";")

In [4]:
head(df)

id age   gender height weight ap_hi ap_lo cholesterol gluc smoke alco active
1 0  18393 2      168    62     110    80   1           1    0     0    1     
2 1  20228 1      156    85     140    90   3           1    0     0    1     
3 2  18857 1      165    64     130    70   3           1    0     0    0     
4 3  17623 2      169    82     150   100   1           1    0     0    1     
5 4  17474 1      156    56     100    60   1           1    0     0    0     
6 8  21914 1      151    67     120    80   2           2    0     0    0     
  cardio
1 0     
2 1     
3 1     
4 1     
5 0     
6 0

**<font color='red'>Вопрос 1.</font> Сколько мужчин и женщин представлено в этом наборе данных? Не было дано расшифровки признака "пол" (какому полу соответствует 1, а какому – 2 в признаке ``gender``) – это определите, посмотрев также на рост при разумном предположении, что в среднем мужчины выше (здесь и далее под средним понимается среднее арифметическое).**

In [5]:
mean(df$height[df$gender == 1])

[1] 161.3556

In [6]:
mean(df$height[df$gender == 2])

[1] 169.9479

- 1 соответствует женщинам, 2 - мужчинам

In [16]:
female <- which((df$gender == 1) == TRUE)
male <- which((df$gender == 2) == TRUE)

In [17]:
length(female)
length(male)

[1] 45530

[1] 24470

***Ответ: 45530 женщин и 24470 мужчин***

**<font color='red'>Вопрос 2.</font> Кто в среднем реже указывает, что употребляет алкоголь – мужчины или женщины?**

In [22]:
length(which((df$alco == 1) & (df$gender == 1)) == TRUE)

[1] 1161

In [20]:
 length(which((df$alco == 1) & (df$gender == 2)) == TRUE)

[1] 2603

***Ответ: женщины в среднем реже указывают, что употребляют алкоголь***

**<font color='red'>Вопрос 3.</font> Во сколько раз (округленно, `round`) процент курящих среди мужчин больше, чем процент курящих среди женщин (по крайней мере, по этим анкетным данным)?**

In [34]:
for_wom <- (length(which((df$smoke == 1) & (df$gender == 1)) == TRUE) / length(female)) 
for_men <- (length(which((df$smoke == 1) & (df$gender == 2)) == TRUE) / length(male))
round(for_men / for_wom)

[1] 12

***Ответ: Процент курящих среди мужчин в 12 раз больше, чем процент курящих среди женщин***

**<font color='red'>Вопрос 4.</font> Вы наверняка заметили, что значения возраста какие-то странные. Догадайтесь, в чём здесь измеряется возраст, и ответьте, на сколько месяцев (примерно) отличаются медианные значения возраста курящих и некурящих.**

In [47]:
df$age_years <- df$age / 365.25
smoke <- median(df$age_years[df$smoke == 1])
nonsmoke <- median(df$age_years[df$smoke == 0])
round((nonsmoke - smoke)*12)

[1] 20

***Ответ: Медианные значения возраста курящих и некрящих отличаются приблизительно на 20 месяцев***

**<font color='red'> Вопрос 5.</font>Интересуют 2 подвыборки курящих мужчин возраста от 60 до 64 лет включительно: первая с верхним артериальным давлением строго меньше 120 мм рт.ст. и концентрацией холестерина – 4 ммоль/л, а вторая – с верхним артериальным давлением от 160 (включительно) до 180 мм рт.ст. (не включительно) и концентрацией холестерина – 8 ммоль/л.
Во сколько раз (округленно, `round`) отличаются доли больных людей (согласно целевому признаку, `cardio`) в этих двух подвыборках? Посчитайте на наших данных.**

In [76]:
men2 <- df$cardio[(df$gender == 2) & (df$smoke == 1) & (df$age_years >= 60) & (df$age_years <= 64) 
                  & (df$ap_hi >= 160) & (df$ap_hi < 180) & (df$cholesterol == 3)]
men1 <- df$cardio[(df$gender == 2) & (df$smoke == 1) & (df$age_years >= 60) & (df$age_years <= 64) 
                  & (df$ap_hi < 120) & (df$cholesterol == 1)]
round(mean(men2)/mean(men1))

[1] 3

***Ответ: доля больных из первой выборки в три раза меньше, чем доля больных из второй***

**<font color='red'>Вопрос 6.</font> Постройте новый признак – BMI ([Body Mass Index](https://en.wikipedia.org/wiki/Body_mass_index)). Для этого надо вес в килограммах поделить на квадрат роста в метрах. Нормальными считаются значения BMI от 18.5 до 25. Выберите верные утверждения.**

**<font color='red'>Утверждения:</font>**
- Медианный BMI по выборке превышает норму
- У женщин в среднем BMI ниже, чем у мужчин
- У здоровых в среднем BMI выше, чем у больных
- В сегменте здоровых и непьющих мужчин в среднем BMI ближе к норме, чем в сегменте здоровых и непьющих женщин

In [77]:
df$BMI <- df$weight / (df$height/100)**2

In [78]:
head(df)

id age   gender height weight ap_hi ap_lo cholesterol gluc smoke alco active
1 0  18393 2      168    62     110    80   1           1    0     0    1     
2 1  20228 1      156    85     140    90   3           1    0     0    1     
3 2  18857 1      165    64     130    70   3           1    0     0    0     
4 3  17623 2      169    82     150   100   1           1    0     0    1     
5 4  17474 1      156    56     100    60   1           1    0     0    0     
6 8  21914 1      151    67     120    80   2           2    0     0    0     
  cardio age_years BMI     
1 0      50.35729  21.96712
2 1      55.38125  34.92768
3 1      51.62765  23.50781
4 1      48.24914  28.71048
5 0      47.84120  23.01118
6 0      59.99726  29.38468

In [81]:
# Утверждение 1 - верно
median(df$BMI)

[1] 26.37407

In [82]:
# Утверждение 2 - неверно
mean(df$BMI[df$gender == 1])
mean(df$BMI[df$gender == 2])

[1] 27.98758

[1] 26.75444

In [83]:
# Утверждение 3 - неверно
mean(df$BMI[df$cardio == 0])
mean(df$BMI[df$cardio == 1])

[1] 26.54818

[1] 28.56606

In [85]:
# Утверждение 4 - верно
mean(df$BMI[(df$cardio == 0) & (df$alco == 0) & (df$gender == 2)])
mean(df$BMI[(df$cardio == 0) & (df$alco == 0) & (df$gender == 1)])

[1] 25.87264

[1] 26.84541

***Ответ: верны первое и последнее утверждения***

**<font color='red'>Вопрос 7.</font> ** 

Отфильтруйте следующие сегменты пациентов (считаем это ошибками в данных)
- указанное нижнее значение артериального давления строго выше верхнего
- рост строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили 
- вес строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили
Сколько процентов данных (округленно, `round`) мы выбросили?

In [112]:
filtered_data <- length(df$BMI[(df$ap_hi >= df$ap_lo) & 
                    (df$height >= quantile(df$height, 0.025)) & (df$height <= quantile(df$height, 0.975)) &
                   (df$weight >= quantile(df$weight, 0.025)) & (df$weight <= quantile(df$weight, 0.975))])
round(100 - filtered_data * 100 / length(df$BMI))

[1] 10

***Ответ: выбросили 10 процентов исходных данных***